src/serac/physics/materials/thermal_functional_material.hpp

```diff
@@ -0,0 +1,239 @@
+ *
+ * @tparam dim Spatial dimension
+ */
+template <int dim>
+struct LinearConductor {
+  /// Density
+  double density_ = -1.0;
+
+  /// Specific heat capacity
+  double specific_heat_capacity_ = -1.0;
+
+  /// Constant isotropic thermal conductivity
+  tensor<double, dim, dim> conductivity_ = {-1.0};
+
+  /**
+   * @brief Function defining the thermal flux (constitutive response)
+   *
+   * @tparam T1 type of the temperature (e.g. tensor or dual type)
+   * @tparam T2 type of the temperature gradient (e.g. tensor or dual type)
+   * @param temperature_gradient Gradient of the temperature (du_dx)
+   * @return The thermal flux of the material model
+   */
+  template <typename T1, typename T2>
+  SERAC_HOST_DEVICE auto operator()(T1& /* temperature */, T2& temperature_gradient) const
+  {
+    for (int i = 0; i < dim; ++i) {
+      SLIC_ASSERT_MSG(conductivity_(i, i) > 0.0,
+                      "Conductivity tensor must be positive definite in linear conductor material model.");
+    }
+    return -1.0 * conductivity_ * temperature_gradient;
+  }
+
+  /**
+   * @brief The density (mass per volume) of the material model
+   *
+   * @tparam T1 type of the position variable
+   * @return The density
+   */
+  template <typename T1>
+  SERAC_HOST_DEVICE double density(T1& /* x */) const
+  {
+    SLIC_ASSERT_MSG(density_ > 0.0, "Density must be positive in the linear conductor material model.");
+
+    return density_;
+  }
+
+  /**
+   * @brief The specific heat capacity (heat capacity per unit mass) of the material model
+   *
+   * @tparam T1 Type of the position variable
+   * @tparam T2 Type of the temperature variable
+   */
+  template <typename T1, typename T2>
+  SERAC_HOST_DEVICE double specificHeatCapacity(T1& /* x */, T2& /* temperature */) const
+  {
+    SLIC_ASSERT_MSG(specific_heat_capacity_ > 0.0,
+                    "Specific heat capacity must be positive in the linear conductor material model.");
+
+    return specific_heat_capacity_;
+  }
+};
+
+// Use SFINAE to add static assertions checking if the given thermal material type is acceptable
+template <typename T, int dim, typename = void>
+struct has_density : std::false_type {
+};
+
+template <typename T, int dim>
+struct has_density<T, dim, std::void_t<decltype(std::declval<T&>().density(std::declval<tensor<double, dim>&>()))>>
+    : std::true_type {
+};
+
+template <typename T, int dim, typename = void>
+struct has_specific_heat_capacity : std::false_type {
+};
+
+template <typename T, int dim>
+struct has_specific_heat_capacity<T, dim,
+                                  std::void_t<decltype(std::declval<T&>().specificHeatCapacity(
+                                      std::declval<tensor<double, dim>&>(), std::declval<tensor<double, 1>&>()))>>
+    : std::true_type {
+};
+
+template <typename T, int dim, typename = void>
+struct has_thermal_flux : std::false_type {
+};
+
+template <typename T, int dim>
+struct has_thermal_flux<
```

Comment:
These checks are useful for checking the existence of an `operator()` that accepts certain arguments, but it's worth noting that there are invalid material definitions that will pass this test, and there are valid material definitions that will not pass it (although, this case is unlikely to happen in practice).

Reply:
What is a valid material that wouldn't pass? I mostly added these checks for human-readable compiler output if a user defined material is ill-formed.

Reply:
the actual arguments that get passed to a q-function will be tensors of dual numbers, so you could explicitly write out the actual types to be passed in (I don't recommend this, as it is verbose and prevents the material from being composable in multiphysics simulations):

```cpp
struct ThisMaterialWorksWithFunctionalButFailsCheck {
  ...
  auto operator()(dual < double > u, tensor< dual < tensor < double, 3 > >, 3 > du_dx) {
    // implement the heat flux calculation
  }
};
```